entity-service-impl/src/main/java/org/hypertrace/entity/query/service/EntityQueryServiceImpl.java

```diff
@@ -289,20 +289,127 @@ private void doUpdate(RequestContext requestContext, EntityUpdateRequest request
+    } catch (Exception e) {
+      responseObserver.onError(
+          new ServiceException("Error occurred while executing " + request, e));
+    }
+  }
+
+  private List<Entity> getProjectedEntities(
+      Iterable<String> entityIdsList,
+      List<Expression> selectionList,
+      RequestContext requestContext) {
+    Query entitiesQuery = Query.newBuilder().addAllEntityId(entityIdsList).build();
+    List<String> docStoreSelections =
+        entityQueryConverter.convertSelectionsToDocStoreSelections(requestContext, selectionList);
+    Iterator<Document> documentIterator =
+        entitiesCollection.search(
+            DocStoreConverter.transform(
+                requestContext.getTenantId().orElseThrow(), entitiesQuery, docStoreSelections));
+    return convertDocsToEntities(documentIterator);
+  }
+
+  private void doBulkUpdate(
+      RequestContext requestContext, Map<String, EntityUpdateInfo> entitiesMap) throws Exception {
+    Map<Key, Map<String, Document>> entitiesUpdateMap = new HashMap<>();
+    for (String entityId : entitiesMap.keySet()) {
+      Map<String, Document> transformedUpdateOperations =
+          transformUpdateOperations(
+              entitiesMap.get(entityId).getUpdateOperationList(), requestContext);
+      if (transformedUpdateOperations.isEmpty()) {
+        continue;
+      }
+      entitiesUpdateMap.put(
+          new SingleValueKey(requestContext.getTenantId().orElseThrow(), entityId),
+          transformedUpdateOperations);
+    }
+
+    if (entitiesUpdateMap.isEmpty()) {
+      LOG.error("There are no entities to update!");
+      return;
+    }
+
+    try {
+      entitiesCollection.bulkUpdateSubDocs(entitiesUpdateMap);
+    } catch (Exception e) {
+      LOG.error("Failed to update entities {}", entitiesMap, e);
+      throw e;
+    }
+  }
+
+  private Map<String, Document> transformUpdateOperations(
+      List<UpdateOperation> updateOperationList, RequestContext requestContext) {
+    Map<String, Document> documentMap = new HashMap<>();
+    for (UpdateOperation updateOperation : updateOperationList) {
+      if (!updateOperation.hasSetAttribute()) {
+        continue;
+      }
+      SetAttribute setAttribute = updateOperation.getSetAttribute();
+      String attributeId = setAttribute.getAttribute().getColumnName();
+      String subDocPath =
+          entityAttributeMapping
+              .getDocStorePathByAttributeId(requestContext, attributeId)
+              .orElseThrow(
+                  () -> new IllegalArgumentException("Unknown attribute FQN " + attributeId));
+      AttributeValue attributeValue =
+          EntityQueryConverter.convertToAttributeValue(setAttribute.getValue()).build();
+      try {
+        String jsonValue = DocStoreJsonFormat.printer().print(attributeValue);
+        documentMap.put(subDocPath, new JSONDocument(jsonValue));
+      } catch (Exception e) {
+        LOG.error("Failed to put update corresponding to {} in the documentMap", subDocPath, e);
```

Comment:
Please throw the exception back https://github.com/hypertrace/entity-service/pull/121#discussion_r685744920